File: frontend/src/features/collaboration/utils/cursorTransform.ts
import type {
  CollaborationCursorState,
  CollaborationTextOperation,
} from '../types/collaboration.types';
import { codePointLength } from './operationTransform';

/**
 * Transforms one cursor position after an insert/delete operation.
 *
 * Inserts at or before the cursor shift it right. Deletes before the cursor
 * shift it left, and cursors inside the deleted range collapse to the delete
 * start.
 *
 * @param position - Cursor position in code points.
 * @param op - Applied operation.
 * @param documentLength - Optional post-operation document length for clamping.
 * @returns Transformed cursor position.
 */
export function transformCursorPosition(
  position: number,
  op: CollaborationTextOperation,
  documentLength?: number,
): number {
  let nextPosition = Math.max(0, position);

  if (op.type === 'insert') {
    if (op.pos <= nextPosition) {
      nextPosition += codePointLength(op.text);
    }
  } else if (op.pos < nextPosition) {
    if (nextPosition <= op.pos + op.len) {
      nextPosition = op.pos;
    } else {
      nextPosition = Math.max(op.pos, nextPosition - op.len);
    }
  }

  if (typeof documentLength !== 'number') {
    return nextPosition;
  }

  return clamp(nextPosition, 0, documentLength);
}

/**
 * Transforms and normalizes a selection range after an operation.
 *
 * @param selectionStart - Selection start in code points.
 * @param selectionEnd - Selection end in code points.
 * @param op - Applied operation.
 * @param documentLength - Optional post-operation document length for clamping.
 * @returns Ordered `[start, end]` selection range.
 */
export function transformSelectionRange(
  selectionStart: number,
  selectionEnd: number,
  op: CollaborationTextOperation,
  documentLength?: number,
): readonly [number, number] {
  const start = transformCursorPosition(selectionStart, op, documentLength);
  const end = transformCursorPosition(selectionEnd, op, documentLength);
  return [Math.min(start, end), Math.max(start, end)] as const;
}

/**
 * Transforms a remote cursor state after an applied operation.
 *
 * @param cursor - Remote cursor state.
 * @param op - Applied operation.
 * @param documentLength - Optional post-operation document length for clamping.
 * @returns Cursor state in the post-operation coordinate space.
 */
export function transformCursor(
  cursor: CollaborationCursorState,
  op: CollaborationTextOperation,
  documentLength?: number,
): CollaborationCursorState {
  const [selectionStart, selectionEnd] = transformSelectionRange(
    cursor.selection_start,
    cursor.selection_end,
    op,
    documentLength,
  );

  return {
    ...cursor,
    pos: transformCursorPosition(cursor.pos, op, documentLength),
    selection_end: selectionEnd,
    selection_start: selectionStart,
  };
}

/**
 * Clamps a number to an inclusive range.
 *
 * @param value - Number to clamp.
 * @param min - Inclusive minimum.
 * @param max - Inclusive maximum.
 * @returns Clamped number.
 */
function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(value, max));
}

File: backend-collab/app/domain/collaboration/cursor_transform.py
"""Cursor and selection transformation utilities for plain-text OT.

Remote cursors are not document operations, but they live in the same coordinate
space. Applying the same position-shifting rules keeps presence markers stable
when inserts and deletes arrive concurrently with cursor updates.
"""

from __future__ import annotations

from app.domain.collaboration.operations import (
    TextOperation,
    clamp_index,
    op_delete_len,
    op_text_len,
)
from app.domain.collaboration.schemas import CursorState


def transform_cursor_position(
    position: int,
    op: TextOperation,
    *,
    document_length: int | None = None,
) -> int:
    """Transform one cursor position after an insert or delete operation.

    Example:
        A cursor at position ``5`` becomes ``8`` after inserting ``"abc"`` at
        position ``2``. A cursor inside a deleted range collapses to the delete
        start. The optional ``document_length`` clamps the result to the current
        post-operation document length.
    """

    next_position = max(0, position)

    if op["type"] == "insert":
        insert_pos = int(op["pos"])
        if insert_pos <= next_position:
            next_position += op_text_len(op)
    else:
        delete_pos = int(op["pos"])
        delete_len = op_delete_len(op)
        delete_end = delete_pos + delete_len

        if delete_pos >= next_position:
            next_position = next_position
        elif next_position <= delete_end:
            next_position = delete_pos
        else:
            next_position = max(delete_pos, next_position - delete_len)

    if document_length is None:
        return next_position

    return clamp_index(next_position, document_length)


def transform_selection_range(
    selection_start: int,
    selection_end: int,
    op: TextOperation,
    *,
    document_length: int | None = None,
) -> tuple[int, int]:
    """Transform and normalize a selection range after an operation."""

    start = transform_cursor_position(selection_start, op, document_length=document_length)
    end = transform_cursor_position(selection_end, op, document_length=document_length)
    return (min(start, end), max(start, end))


def transform_cursor_state(
    cursor: CursorState,
    op: TextOperation,
    *,
    document_length: int | None = None,
) -> CursorState:
    """Return a cursor state shifted into the post-operation document.

    The cursor head and selection anchors are transformed independently and then
    clamped to the supplied document length. The timestamp and identity metadata
    are preserved because this utility only changes coordinates.
    """

    selection_start, selection_end = transform_selection_range(
        cursor.selection_start,
        cursor.selection_end,
        op,
        document_length=document_length,
    )
    return cursor.model_copy(
        update={
            "pos": transform_cursor_position(cursor.pos, op, document_length=document_length),
            "selection_start": selection_start,
            "selection_end": selection_end,
        }
    )
